src/compiler/binder.rs

```diff
@@ -0,0 +1,194 @@
+        *self.parent.write().unwrap() = parent;
+    }
+
+    fn bind_source_file(&self, f: Rc<Node>) {
+        self.set_file(Some(f.clone()));
+
+        if true {
+            self.bind(Some(self.file()));
+        }
+
+        self.set_file(None);
+        self.set_parent(None);
+    }
+
+    fn bind_container(&self, node: Rc<Node>, container_flags: ContainerFlags) {
+        self.bind_children(node);
+    }
+
+    fn bind_each_functions_first(&self, nodes: &NodeArray) {
+        BinderType::bind_each_callback(nodes, |n| {
+            if n.kind() == SyntaxKind::FunctionDeclaration {
+                self.bind(Some(n))
+            }
+        });
+        BinderType::bind_each_callback(nodes, |n| {
+            if n.kind() != SyntaxKind::FunctionDeclaration {
+                self.bind(Some(n))
+            }
+        });
+    }
+
+    fn bind_each(&self, nodes: &NodeArray) {
+        for_each(nodes, |node, _| {
+            self.bind(Some(node.clone()));
+            Option::<()>::None
+        });
+    }
+
+    fn bind_each_callback<TNodeCallback: FnMut(Rc<Node>)>(
+        nodes: &NodeArray,
+        mut bind_function: TNodeCallback,
+    ) {
+        for_each(nodes, |node, _| {
+            bind_function(node.clone());
+            Option::<()>::None
+        });
+    }
+
+    fn bind_each_child(&self, node: Rc<Node>) {
+        for_each_child(
+            node,
+            |node| self.bind(Some(node)),
+            |nodes| self.bind_each(nodes),
+        );
+    }
+
+    fn bind_children(&self, node: Rc<Node>) {
+        match &*node {
+            Node::Statement(statement) => match statement {
+                Statement::ExpressionStatement(expression_statement) => {
+                    self.bind_expression_statement(expression_statement);
+                }
+                _ => unimplemented!(),
+            },
+            Node::Expression(expression) => match expression {
+                Expression::PrefixUnaryExpression(_) => {
+                    self.bind_prefix_unary_expression_flow(node);
+                }
+                _ => unimplemented!(),
+            },
+            Node::SourceFile(source_file) => {
+                self.bind_each_functions_first(&source_file.statements);
+            }
+            Node::BaseNode(_) => panic!("Didn't expect to bind BaseNode?"),
+        };
+    }
+
+    fn bind_expression_statement(&self, node: &ExpressionStatement) {
+        self.bind(Some(node.expression.clone()));
+    }
+
+    fn bind_prefix_unary_expression_flow(&self, node: Rc<Node>) {
+        if false {
+        } else {
+            self.bind_each_child(node);
+        }
+    }
+
+    fn get_container_flags(&self, node: Rc<Node>) -> ContainerFlags {
+        match node.kind() {
+            SyntaxKind::SourceFile => {
+                return ContainerFlags::IsContainer
+                    | ContainerFlags::IsControlFlowContainer
+                    | ContainerFlags::HasLocals;
+            }
```

Comment:
Because the following `ContainerFlags::None` is a following statement in this function body, so if we don't explicitly `return` here then this would just be the value of this `match` expression but that would get silently discarded and then `ContainerFlags::None` would always get returned